File: internal/cdn/cdn.go
package cdn

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	cfconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/cloudfront"
	cftypes "github.com/aws/aws-sdk-go-v2/service/cloudfront/types"
	"github.com/google/uuid"

	"hakobin/internal/config"
)

type CdnInvalidator struct {
	Type           string // "cloudfront" or "cloudflare"
	DistributionID string // for cloudfront
	ZoneID         string // for cloudflare
	ApiToken       string // for cloudflare
}

func FromEnv() (*CdnInvalidator, error) {
	kind := strings.TrimSpace(os.Getenv("HAKOBIN_CDN_PURGE_TYPE"))
	if kind == "" {
		return nil, nil
	}

	switch strings.ToLower(kind) {
	case "cloudfront":
		distID := strings.TrimSpace(os.Getenv("CLOUDFRONT_DISTRIBUTION_ID"))
		if distID == "" {
			return nil, fmt.Errorf("CLOUDFRONT_DISTRIBUTION_ID environment variable is required")
		}
		return &CdnInvalidator{
			Type:           "cloudfront",
			DistributionID: distID,
		}, nil
	case "cloudflare":
		zoneID := strings.TrimSpace(os.Getenv("CLOUDFLARE_ZONE_ID"))
		if zoneID == "" {
			return nil, fmt.Errorf("CLOUDFLARE_ZONE_ID environment variable is required")
		}
		apiToken := strings.TrimSpace(os.Getenv("CLOUDFLARE_API_TOKEN"))
		if apiToken == "" {
			return nil, fmt.Errorf("CLOUDFLARE_API_TOKEN environment variable is required")
		}
		return &CdnInvalidator{
			Type:     "cloudflare",
			ZoneID:   zoneID,
			ApiToken: apiToken,
		}, nil
	default:
		return nil, fmt.Errorf("unsupported CDN type: %s", kind)
	}
}

func (c *CdnInvalidator) Invalidate(ctx context.Context, cfg *config.Config, paths []string) error {
	if len(paths) == 0 {
		return nil
	}

	switch c.Type {
	case "cloudfront":
		return c.invalidateCloudFront(ctx, cfg, paths)
	case "cloudflare":
		return c.invalidateCloudflare(ctx, cfg, paths)
	default:
		return fmt.Errorf("unsupported CDN type: %s", c.Type)
	}
}

func (c *CdnInvalidator) invalidateCloudFront(ctx context.Context, cfg *config.Config, paths []string) error {
	// Reuse the same S3 credentials/region the rest of the tool is configured
	// with, rather than silently falling back to the default AWS chain (which
	// breaks when S3 credentials are supplied only via env for a non-default
	// profile or a MinIO-style setup).
	opts := []func(*cfconfig.LoadOptions) error{}
	if cfg != nil {
		if cfg.S3Region != "" {
			opts = append(opts, cfconfig.WithRegion(cfg.S3Region))
		}
		if cfg.S3AccessKeyID != "" && cfg.S3SecretAccessKey != "" {
			opts = append(opts, cfconfig.WithCredentialsProvider(
				credentials.NewStaticCredentialsProvider(cfg.S3AccessKeyID, cfg.S3SecretAccessKey, ""),
			))
		}
	}

	awsCfg, err := cfconfig.LoadDefaultConfig(ctx, opts...)
	if err != nil {
		return fmt.Errorf("failed to load CloudFront config: %w", err)
	}

	client := cloudfront.NewFromConfig(awsCfg)

	var items []string
	for _, p := range paths {
		items = append(items, normalizePath(p))
	}

	// A unique caller reference per invalidation: CloudFront treats a repeated
	// caller reference idempotently, so a second-resolution timestamp could make
	// distinct invalidations in the same second collide and be dropped.
	callerRef := fmt.Sprintf("hakobin-%s", uuid.NewString())

	input := &cloudfront.CreateInvalidationInput{
		DistributionId: aws.String(c.DistributionID),
		InvalidationBatch: &cftypes.InvalidationBatch{
			CallerReference: aws.String(callerRef),
			Paths: &cftypes.Paths{
				Quantity: aws.Int32(int32(len(items))),
				Items:    items,
			},
		},
	}

	result, err := client.CreateInvalidation(ctx, input)
	if err != nil {
		return fmt.Errorf("failed to create CloudFront invalidation: %w", err)
	}

	if result.Invalidation != nil && result.Invalidation.Id != nil {
		fmt.Printf("CloudFront invalidation created: %s\n", *result.Invalidation.Id)
	}
	fmt.Printf("Invalidating %d paths\n", len(items))
	return nil
}

func (c *CdnInvalidator) invalidateCloudflare(ctx context.Context, cfg *config.Config, paths []string) error {
	baseURL, err := cfg.RepositoryBaseURL()
	if err != nil {
		return err
	}
	baseURL = strings.TrimSuffix(baseURL, "/")

	var urls []string
	for _, p := range paths {
		urls = append(urls, fmt.Sprintf("%s%s", baseURL, normalizePath(p)))
	}

	client := &http.Client{Timeout: 10 * time.Second}

	// Cloudflare API allows purging in batches of up to 30 files
	for i := 0; i < len(urls); i += 30 {
		end := i + 30
		if end > len(urls) {
			end = len(urls)
		}
		batch := urls[i:end]

		reqBody, err := json.Marshal(map[string]interface{}{
			"files": batch,
		})
		if err != nil {
			return fmt.Errorf("failed to marshal Cloudflare payload: %w", err)
		}

		reqURL := fmt.Sprintf("https://api.cloudflare.com/client/v4/zones/%s/purge_cache", c.ZoneID)
		req, err := http.NewRequestWithContext(ctx, "POST", reqURL, bytes.NewReader(reqBody))
		if err != nil {
			return fmt.Errorf("failed to create Cloudflare request: %w", err)
		}

		req.Header.Set("Authorization", "Bearer "+c.ApiToken)
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			return fmt.Errorf("failed to send Cloudflare purge request: %w", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			body, _ := ioReadAll(resp.Body)
			return fmt.Errorf("cloudflare purge failed with status %d: %s", resp.StatusCode, string(body))
		}
	}

	fmt.Printf("Cloudflare cache purged for %d URLs\n", len(urls))
	return nil
}

func normalizePath(path string) string {
	if strings.HasPrefix(path, "/") {
		return path
	}
	return "/" + path
}

func ioReadAll(r io.Reader) ([]byte, error) {
	var buf bytes.Buffer
	_, err := io.Copy(&buf, r)
	return buf.Bytes(), err
}

File: src/cdn.rs
use std::time::{SystemTime, UNIX_EPOCH};

use anyhow::Context;
use aws_config::BehaviorVersion;
use aws_sdk_cloudfront::types::{InvalidationBatch, Paths};
use reqwest::StatusCode;
use serde_json::json;

use crate::config::Config;

#[derive(Debug, Clone)]
pub enum CdnInvalidator {
    CloudFront { distribution_id: String },
    Cloudflare { zone_id: String, api_token: String },
}

impl CdnInvalidator {
    pub fn from_env() -> anyhow::Result<Option<Self>> {
        let Some(kind) = std::env::var("HAKOBIN_CDN_PURGE_TYPE")
            .ok()
            .filter(|v| !v.trim().is_empty())
        else {
            return Ok(None);
        };

        match kind.to_lowercase().as_str() {
            "cloudfront" => Ok(Some(Self::CloudFront {
                distribution_id: required_env("CLOUDFRONT_DISTRIBUTION_ID")?,
            })),
            "cloudflare" => Ok(Some(Self::Cloudflare {
                zone_id: required_env("CLOUDFLARE_ZONE_ID")?,
                api_token: required_env("CLOUDFLARE_API_TOKEN")?,
            })),
            other => anyhow::bail!("unsupported CDN type: {other}"),
        }
    }

    pub async fn invalidate(&self, cfg: &Config, paths: &[String]) -> anyhow::Result<()> {
        if paths.is_empty() {
            return Ok(());
        }

        match self {
            Self::CloudFront { distribution_id } => {
                invalidate_cloudfront(distribution_id, paths).await
            }
            Self::Cloudflare { zone_id, api_token } => {
                invalidate_cloudflare(cfg, zone_id, api_token, paths).await
            }
        }
    }
}

async fn invalidate_cloudfront(distribution_id: &str, paths: &[String]) -> anyhow::Result<()> {
    let cfg = aws_config::defaults(BehaviorVersion::latest()).load().await;
    let client = aws_sdk_cloudfront::Client::new(&cfg);
    let normalized = paths.iter().map(normalize_path).collect::<Vec<_>>();
    let quantity = normalized.len() as i32;
    let caller_reference = format!(
        "hakobin-{}",
        SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs()
    );

    let batch = InvalidationBatch::builder()
        .caller_reference(caller_reference)
        .paths(
            Paths::builder()
                .quantity(quantity)
                .set_items(Some(normalized.clone()))
                .build()?,
        )
        .build()?;

    let result = client
        .create_invalidation()
        .distribution_id(distribution_id)
        .invalidation_batch(batch)
        .send()
        .await
        .context("failed to create CloudFront invalidation")?;

    if let Some(invalidation) = result.invalidation {
        println!("CloudFront invalidation created: {}", invalidation.id);
    }
    println!("Invalidating {} paths", normalized.len());
    Ok(())
}

async fn invalidate_cloudflare(
    cfg: &Config,
    zone_id: &str,
    api_token: &str,
    paths: &[String],
) -> anyhow::Result<()> {
    let base_url = cfg.repository_base_url()?;
    let urls = paths
        .iter()
        .map(|path| format!("{}{}", base_url.trim_end_matches('/'), normalize_path(path)))
        .collect::<Vec<_>>();

    let client = reqwest::Client::new();
    for batch in urls.chunks(30) {
        let response = client
            .post(format!(
                "https://api.cloudflare.com/client/v4/zones/{zone_id}/purge_cache"
            ))
            .bearer_auth(api_token)
            .json(&json!({ "files": batch }))
            .send()
            .await
            .context("failed to send Cloudflare purge request")?;

        if response.status() != StatusCode::OK {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            anyhow::bail!("Cloudflare purge failed with {status}: {body}");
        }
    }

    println!("Cloudflare cache purged for {} URLs", urls.len());
    Ok(())
}

fn normalize_path(path: &String) -> String {
    if path.starts_with('/') {
        path.clone()
    } else {
        format!("/{path}")
    }
}

fn required_env(name: &str) -> anyhow::Result<String> {
    std::env::var(name)
        .ok()
        .filter(|v| !v.trim().is_empty())
        .with_context(|| format!("{name} environment variable is required"))
}
